### PR TITLE
Setting the artist page to read from Statebus

### PR DESCRIPTION
**Pull request by cfix-git**
Updated:

The artists entry form has new keys: tagline and website. 
The artists page is mostly the same. 

There is one issue:

I can not determine why each input is printing {entity.key}+{entity.value} to the page. That needs to go away.

### CODE REVIEW

File: residency/artists-enter.html [hunk 1]
```diff
@@ -33,44 +33,47 @@
...
+                className: "read-more"
+                dataName: artist.name
+                onClick: toggleDescription
+                "More"
+              P
+                className: "artist-description"
+                id: artist.name + "-descr"
+                artist.description
+                if artist.website
+                  A href: artist.website
```
Comment (learner-long-life) [inline residency/artists-enter.html:82]: nice, i didn't know you could do this, for a single attribute.

**Comment by toomim** [comment]
It's cool seeing this statebus client code. Cortland, how has it been to write it? I'd love to know what issues you've run into and how I can help.

**Comment by toomim** [comment]
A good place to start might be helping you with the printing of entity.key + entity.value. What specifically are you referring to? I've loaded https://invisible.college/residency/artists-enter ... is there something on that page that shouldn't be there?

**Comment by cfix-git** [comment]
At [https://invisible.college/residency/artists-enter](https://invisible.college/residency/artists-enter) near the bottom, the following string is printed: "nameNamephotophotodescriptionDescription"

Which is each key/value pair representing the inputs. 

Also, many of my original functions are still written in plain JS cuz i didn't want to do CoffeeScript. They seem to call well enough. 

I like they style of syntax Statebus offers compared to HTML. It is nice to see a data oriented approach to writing a document. It'll take a little getting used to, but I think it can be powerful. 

An issue I ran into was checking when the content was loaded. Specifically, the photos should animate only after they have been visible on the page. A `window.onload` does not call the function, but setTimeout(animatePhotos, 5) yields the desired result. In fact, this shows nothing in the console:

```
window.onload = ->
      console.log("loaded")
```
Any insight?

**Comment by toomim** [comment]
> At https://invisible.college/residency/artists-enter near the bottom, the following string is printed: "nameNamephotophotodescriptionDescription"

I see the problem. When you construct HTML, you can't include imperative lines in the middle of an element:

```coffeescript
dom.BODY = ->
  DIV
    style:
      textAlign: 'left'
      padding: '20px'
      marginBottom: '50px'
      font: "18px 'Muli', sans-serif"

    artists_list = fetch("/residency/artists") or {}   # This isn't good

    if not artists_list.key?                           # This either
      artists_list.key = "/residency/artists"
    if not artists_list.list?                          # This either
      artists_list.list = []
```

The `fetch("/residency/artists")` line returns `{"key":"/residency/artists","list":[{"name":"Latasha Alcindor", ...},{"key":"/ls/aip", ... }],"lastId":null}`, which is interpreted as a set of attributes for the DIV, and thus it makes a "list" attribute and sets it to "[object Object],[object Object]" here in the DOM  output:

```html
<div style="text-align:left;padding:20px;margin-bottom:50px;font:18px 'Muli', sans-serif;" list="[object Object],[object Object]" data-widget="BODY" data-reactid=".0">
```

I think the nameNamephotoPhotodescriptionDescription stuff is coming from the line down at the bottom, which is also being included in the BODY DIV:
```coffeescript
    aip_entries = [
        {key: "name", name: "Name"},
        {key: "photo", name: "photo"}
        {key: "description", name: "Description"}
    ]
```

The solution is to do all this imperative assignment stuff at the *top* of the function, before you return the DIV. Inside the DIV, it's ok to have if statements that evaluate to null, undefined, or false (because any such value will be ignored) but you can't have any imperative statement that evaluates to something (like these) unless you want that something to actually get included in the HTML that's output by the function.

I know this is a bit confusing. I hope we can make it better.

**Comment by toomim** [comment]
Javascript is totally fine. I only use the coffeescript because it's a sweet way to define nested HTML. The javascript equivalent is:

```javascript
dom.BODY = () => {
  if (this.loading())
    return "Loading..."

  return DIV(
    {
      style: {
        textAlign: 'left',
        padding: '20px',
        marginBottom: '50px',
        font: "18px 'Muli', sans-serif",
      }
    },

    DIV(
      {
        style: {
          fontWeight: 'bold',
          fontSize: '1.5em',
        }
      },
      'Enter the Artmaker'
    )
  )
```

It's nice to be able to skip all those curly braces and parentheses.

Greg Little has actually been developing a Javascript preprocessor that borrows *only* this one nicety of coffeescript for generating nested HTML. He's calling it decalf. It mostly works, but I think we can improve the syntax:
```
DIV ~
  width : 50
  height : ~big ? 3 : 20
  SPAN ~
    style : ~
      backgroundColor : 'red'
    "hello"
```
Right now it requires a tilde ~ before each additional nesting level.

There's a demo at https://glittle.org/aa/decaf2 -- press control-enter to run the code and see its output.

**Comment by toomim** [comment]
Side note—this doesn't work:
```coffeescript
dom.BODY = ->
  if (@loading)
    "Loading..."
  DIV
    ...
```

First of all, `@loading` needs to be `@loading()` or else it won't call the function, it'll just return it. And that function always exists, and thus if (@loading) will always return true.

Second, you don't need the parens around @loading. It can be `if @loading()`.

Third, it isn't doing anything, because there's no "return" statement. Coffeescript inserts an explicit return statement for the last expression in a function, but the last expression in this function is the DIV at the bottom, so the "Loading..." text just does nothing.

Fourth, it so happens that the statebus system automatically displays a "Loading..." indicator if there's any sort of problem while loading. So it'll probably show "Loading..." already, and if not, you can make it work with this:
```coffeescript
if @loading()
  throw 'barf'
```
and if desired, you can do a one-liner if statement you add a `then` like `if @loading() then throw 'barf'`.

**Comment by cfix-git** [comment]
Thanks @toomim. I've updated the changes, and it should be fine for a merge.

**Comment by learner-long-life** [comment]
Hey, the imperative statements rendering and `@loading` were because that was one of my first attempts to write Statebus code :)
It looks great and I'll merge this PR.
Awesome first commit @ckmahoney

**Comment by learner-long-life** [comment]
@ckmahoney I added you as an admin to this repo so you can merge your own PRs in the future.
They're useful to get feedback, but it's up to your judgment.
For small changes or things you don't need feedback on, you can go ahead and merge / commit to master.

Any of us can fix it or roll it back if it causes problems.

**Comment by cfix-git** [comment]
Neato! Thanks @cryptogoth 

I will certainly reach out on items and issues that need a second set of eyes. Glad to see this one going through well.

**Pull request merged**